Code

jagotu opened the pull request.

jagotu committed:
inital commit

jagotu committed:
signatures

jagotu committed:
methods

jagotu committed:
The tree is now compilable!

jagotu committed:
Small optimisations
from 4-5x times slower than dotnet to 1.75-2x

jagotu committed:
Node referenced by bytecode patching

jagotu committed:
Developement progress:
* foreign method calls with dependent assembly loading
* ManagedReferences
* Table loading now actually compliant with big tables

jagotu committed:
References are now actually references

jagotu committed:
Fix prepareArgs bug

jagotu committed:
Add language launcher

jagotu committed:
Basic type system and classes support

jagotu committed:
Small fixes

jagotu committed:
(unfinished) locations optimalization

jagotu committed:
another different way of storing locations
* long[] instead of byte[] accessed with unsafe
* temporarily deleted narrowing and widening to find raw performance issues

jagotu committed:
Small enhancements:
* pop opcode support
* function call fix

jagotu committed:
Better valuetypes support, indirect load/store

jagotu committed:
Internal components/types/methods support

BACILHelpers + BACILConsoleType + BACILConsoleWriteMethod = stdout

jagotu committed:
TickCount

jagotu committed:
Static fields

jagotu committed:
Remove debugging profanity

jagotu committed:
Call init when accessing static fields

jagotu committed:
More intructions, better timer

jagotu committed:
Bugfix

jagotu committed:
Bugfix

jagotu committed:
Modules, manifests

jagotu committed:
Arrays

jagotu committed:
Optimizations

Creating a bogus 0-element array is better optimized than possibly having null pointers

jagotu committed:
Nodeize BOX and LDSTR

This allows caching of type for BOX (type lookup should never be on compiled path) and caching of the string value for LDSTR.

jagotu committed:
Float & double implementation

Also added CONV opcodes

jagotu committed:
InternalCall methods support

Also added neg and ldlen instructions

jagotu committed:
Unsigned/Unordered comparisons

jagotu committed:
Basic fields inheritance

jagotu committed:
Mostly smaller fixes

Changes driven by dotnet's CodeGenBringUpTests (https://github.com/dotnet/runtime/tree/v5.0.6/src/tests/JIT/CodeGenBringUpTests)

jagotu committed:
Virtual methods implemented

jagotu committed:
Added README.md

jagotu committed:
Fix warnings

* use UnsupportedOperationException instead of (sun internal) NotImplementedException
* remove unused sun.misc.Unsafe wrappers
* update junit dependency (Dependabot security report)

jagotu committed:
Delete unused NativePointer

Forgot to remove in previous commit

jagotu committed:
Start working on the text

jagotu committed:
VirtualCallNode: use rootMethod for signature

jagotu committed:
Add parametrized method stubbing

Useful for running code using Console.WriteLine (--cil.stubMethods=WriteLine)

jagotu committed:
Ldtoken + better signature comparison

The old byte[] comparison couldn't work with cross-component class references.

jagotu committed:
InitializeArray internal method support

jagotu committed:
TypeSpec support

jagotu committed:
Math.cos added

jagotu committed:
Fix stubbing methods with `this`

jagotu committed:
Update README.md with link to BACIL_examples

Resolves #2

jagotu committed:
Update README.md

Parentheses blunder in link

jagotu committed:
Documentation++, refactors

jagotu committed:
Documentation++

Also replaced references to Execution stack with Evaluation stack as that's what ECMA-335 calls it.

jagotu committed:
Documentation+++, small refactors

jagotu committed:
More documentation, refactors

jagotu committed:
Create LICENSE

jagotu committed:
Make querying parser tables actually a compilation constant

I designed the parser so that simple queries would get partially-evaluated to a constant, but it actually wasn't happening.
As always, sprinkling a few annotations throughout CompilationFinal around the source was enough to alleviate that.

Currently, no hot paths query the parser, so this should have no performance impact.

jagotu committed:
Fix a bug when using .NET 6

We expected all heap streams to exist, but ECMA-335 II.24.2.2 says "Streams need not be there if they are empty."

jagotu committed:
Thesis text: Designing a CLI Component parser

Comes with 1 additional CompilationFinal directive for free :)

jagotu committed:
Fix .cctor signature

For non-generic methods we set the GenParamCount to -1, while the signature was looking for a 0.

jagotu committed:
Remove unnecessary table size check

Turns out we actually support tables with more than 65535 rows since 0c8b9914bfe3030e842f85bbf3be020a5fb1e857, which added penalties even for non-coded indices.

jagotu committed:
Thesis text: Parser chapter is content complete

jagotu committed:
Thesis text: structure, parts of introduction

jagotu committed:
Thesis text: example of compiler graphs

for discussion

jagotu committed:
Actually include the graphs, oops

jagotu committed:
Thesis text: .NET/CLI

jagotu committed:
Thesis text: Partial evaluation theory

jagotu committed:
Thesis text: truffle and graal

jagotu committed:
Thesis text: Introduction grammar

jagotu committed:
Thesis text: Theory

jagotu committed:
Thesis text: Relevance of Futamura projections

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+# Abstract
+
+Interpreted and just-in-time (JIT) compiled languages are becoming more and more prevalent, with JavaScript and Python ranking in TOP 5 most popular programming languages in most surveys and statistics. In the academic and research settings, the ability to quickly prototype programming languages and test their properties also works in favor of interpreted languages, as implementing an interpreter ...
```
Není mi jasné, co se tím chce říct: "Interpreted and just-in-time (JIT) compiled languages are becoming more and more prevalent" -- vs. co? AOT compiled languages? Dále: "works in favor of interpreted languages" -- takže nás zajimají interpreter languages? Klasický nitpick koholiv z prostředí PL research by byl, že neexistují interpretované ani kompilované jazyky, ale jejich implementace.

Obecně tahle kapitola míří dobrým směrem: zdůvodnit, proč mělo smysl to celé dělat -- že chceme zjistit, jestli to půjde s Trufflem rychleji a/nebo to bude v něčem lepší než tradiční přístup.

jagotu replied:
Tohle je vykopírovaný abstract ze zadání práce který už je dlouho zapsaný v SISu, s ním už nic neudělám :)

steve-s replied:
Ah, ja jsem si rikal. Njn

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+While writing an interpreter for even a fairly complicated language is achievable for a single person interested in the topic, state-of-the-art optimizing compilers are usually created over several years by large teams of developers at the largest IT companies. Not only was kickstarting such a project unthinkable for an individual but even introducing changes to an already existing project is far ...
+
+For example, Google's state-of-the-art JavaScript engine V8 currently has two different JIT compilers and its own internal bytecode. An experiment of adding a single new bytecode instruction to the project can mean several days of just orientating in the codebase.
```
minor: místo currently bych používal všude "as of 2022" nebo podobné

steve-s replied:
a/nebo citace by byla pekna, oponenti maji radi citace

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+For example, Google's state-of-the-art JavaScript engine V8 currently has two different JIT compilers and its own internal bytecode. An experiment of adding a single new bytecode instruction to the project can mean several days of just orientating in the codebase.
+
+As cybersecurity becomes a more important topic, another factor to consider is that creating manual optimizations in JITs is very prone to bugs which can have very grave security implications. Speculated assumptions of JIT compilers introduced whole new bug families including "type confusion". Implementing a JIT that is not only performant but also secure is proving to be difficult even for state ...
```
Na tohle by taky mohla byt nejaka hezka citace

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+To implement a high-performance CLI runtime, we alleviate the 
+ [Truffle language implementation framework](https://www.graalvm.org/graalvm-as-a-platform/language-implementation-framework/) (henceforth "Truffle") and the [GraalVM Compiler](https://www.graalvm.org/22.1/docs/introduction). These two components are tightly coupled together and we'll mostly be referring to them that way. 
+ 
+The Graal Compiler is a general high-performance just-in-time compiler for Java bytecode that is itself written in Java. It is state-of-the-art when it comes to optimization algorithms - according to official documentation, "the compiler in GraalVM Enterprise includes 62 optimization phases, of which 27 are patented". 
```
reference

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+Also, it is often useful to make sure some exceptional code paths are never included in the compilation - for example, if dividing by zero should result in an immediate crash of the application with a message being printed out, there is no use in spending time compiling and optimizing the error-message printing code, as it will at max be called once.
+
+For that, Truffle uses guards - statements that when reached by the runtime result in de-optimization. De-optimization is a process of transferring evaluation from the compiled variant of the method back to the interpreter (at the precise point where it was interrupted) and throwing away the already compiled variant, as its assumptions no longer hold.
```
Timto jsou myslene guards v Graal IR? V Tuffle DSL, ktery jsi, myslim, nepouzil je take koncept guardu a je to trochu neco jineho. Nevim, jestli to neni matouci.

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+ 
+The Graal Compiler is a general high-performance just-in-time compiler for Java bytecode that is itself written in Java. It is state-of-the-art when it comes to optimization algorithms - according to official documentation, "the compiler in GraalVM Enterprise includes 62 optimization phases, of which 27 are patented". 
+
+Truffle on the other hand is a framework for implementing languages that will run on Graal. From the outside, it behaves like a compiler: its job is to take guest language code and convert it to the VM's language. Unlike a hand-crafted compiler, Truffle takes an interpreter of the guest language as its input and uses [Partial evaluation](#partial-evaluation) to do the compilation, performing a so ...
```
will be compiled by Graal? or will run on GraalVM? Nekdy se rozlisuje Truffle framework nebo API (implementacni framework pro autory interpretru) a Truffle compiler == Graal compiler + compiler pluginy pro partial evaluation, ktere rozumneji tomu Truffle API

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+As long as the virtual call is effectively static at runtime, we only spend time compiling the actual target function (which can be specialized for the environment) and during invocation only pay the price of a simple equality check.
+
+Once the comparison fails, this version of the compiled method is thrown away, and a generic one is created:
```
Mozna by se hodilo vysvetlit v cem je to lepsi nez:
```
if (obj.type == expectedType) return cache.call();
// else no deopt
return obj.Type.ResolveVirtualFunc(method).call();
```
takhle delal .NET profile-guided optimalizace posledne, kdyz jsem se na to dival, protoze jeste neumeli deopt, ale umeli tiered kompilaci

steve-s commented on thesis_text/thesis_text.md:
````diff
@@ -0,0 +1,551 @@
+  return vars[a]
+```
+
+The original control flow of the method was reconstructed from flat bytecode just by unrolling the interpreter loop and merging the instances having the same bytecode offset.
````
Neni mi to ted jasne: "was reconstructed from flat bytecode just by unrolling the interpreter loop" -- jakoze jsme si to ted rucne udelali, abychom pak chapali, co dela Graal/Truffle compiler? Nebo was reconstructed by Truffle compiler? V tom pripade by tu mohl byt skec toho interpreter loopu?

steve-s commented on thesis_text/thesis_text.md:
```diff
@@ -0,0 +1,551 @@
+
+### Design goals
+
+Before we started designing and implementing the parser, we considered what additional constraints have to be put on a parser in order for it to be partial-evaluation friendly. For partial-evaluation friendliness, the key metric is how trivial can every piece of code (that can be called a hot path) be partially evaluated to. While our goal was for the parser to never be called on a hot path, for  ...
```
"hot path" se bezne mezi Truffle vyvojari pouziva, presnejsi je asi neco jako "included in partial evaluation" nebo tak podobne. Mozna bys mohl nekde na zacatku definovat, co pro nas bude znamenat "hot path". IMO performance parseru je taky dulezita, coz je jedna z vyhod tve prace oproti te praci z Lince, a tak i v parseru budou nejake "hot paths" v urcitem slova smyslu

steve-s replied:
Premyslim, jestli to nejak nedostat do problem statementu, protoze v jistem smyslu to byl tvuj goal udelat tohle lip nez to mel ten clovek z Lince. Objektivne vzato, pokud by bylo cilem jen udelat rychly interpreter, pak zkraceni cesty k tomuto cili pres zjenoduseni parsovani by mohlo byt validni. Dalsi hezka vec je, ze mas i language launcher podle Truffle "best practices", treba by se to taky dalo prodat i s tim

jagotu committed:
Thesis: Small improvements

jagotu committed:
Implement NOT instruction

jagotu committed:
Thesis text: Completeness, Benchmark against Hagmüller's work

jagotu committed:
Thesis text: Values and locations

jagotu committed:
I might or might not have just implemented ValueTypes and unbox.any just to get 7 more successful tests

ValueTypes are unrolled (flattened) to their respective LocationsHolder. On evaluation stack they exist as plain LocationHolders. Most changes are caused by the fact that LocationDescriptor has  to remember offsets to both the primitives and refs arrays to fully mark a beginning of ValueType. A GetHashCode internalcall had to be implemented for them.

jagotu committed:
ValueType fixes

jagotu committed:
Call Main with an empty string array instead of null

jagotu committed:
Init ValueTypes before creating an array of them

Required to generate the instanceFieldsDescriptor

jagotu committed:
Thesis: Benchmarks done!

jagotu committed:
Readme: links to thesis chapters

jagotu committed:
Thesis++

jagotu committed:
Thesis: Nodes

jagotu committed:
Thesis: standard libraries, BACILHelpers

jagotu committed:
Thesis: I confused abstract with goal statement (zadání)

jagotu committed:
Thesis: Debugging performance issues

Apart from abstract and conclusion the thesis should be content-complete!

jagotu committed:
Thesis: Tiered compilation, some grammar

jagotu committed:
Thesis: grammar, readability

jagotu committed:
Thesis: update parser size

jagotu committed:
Thesis: more readability improvements

jagotu committed:
Thesis: more readability improvements part 2

jagotu committed:
Thesis: more readability improvements part 3

jagotu committed:
Update README.md

Finally added a link to the thesis
